posts/blundell/You_have_permission_to.jade

```diff
@@ -0,0 +1,204 @@
+    import android.support.v7.app.AppCompatActivity;
+    import android.widget.Toast;
+
+    public class MainActivity extends AppCompatActivity {
+
+        @Override
+        protected void onCreate(Bundle savedInstanceState) {
+            super.onCreate(savedInstanceState);
+            setContentView(R.layout.activity_main);
+
+            if (weHavePermissionToReadContacts()) {
+                readTheContacts();
+            } else {
+                requestReadContactsPermissionFirst();
+            }
+        }
+
+        private boolean weHavePermissionToReadContacts() {
+            return ContextCompat.checkSelfPermission(this, Manifest.permission.READ_CONTACTS) == PackageManager.PERMISSION_GRANTED;
+        }
+
+        private void readTheContacts() {
+            Contacts.getLookupUri(getContentResolver(), Contacts.CONTENT_URI);
+        }
+
+        private void requestReadContactsPermissionFirst() {
+            if (ActivityCompat.shouldShowRequestPermissionRationale(this, Manifest.permission.READ_CONTACTS)) {
+                Toast.makeText(this, "We need permission so you can text your friends.", Toast.LENGTH_LONG).show();
+                requestForResultContactsPermission();
+            } else {
+                requestForResultContactsPermission();
+            }
+        }
+
+        private void requestForResultContactsPermission() {
+            ActivityCompat.requestPermissions(this, new String[]{Manifest.permission.READ_CONTACTS}, 123);
+        }
+
+        @Override
+        public void onRequestPermissionsResult(int requestCode, String[] permissions, int[] grantResults) {
+            super.onRequestPermissionsResult(requestCode, permissions, grantResults);
+            if (requestCode == 123
+                    && grantResults.length > 0
+                    && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
+                Toast.makeText(this, "Permission Granted", Toast.LENGTH_SHORT).show();
+                readTheContacts();
+            } else {
+                Toast.makeText(this, "Permission Denied", Toast.LENGTH_SHORT).show();
+            }
+        }
+    }
+    ```
+
+    `AndroidManfest.xml` stays the same.
+
+    I have extracted methods in this code to try and make it as self explanatory as possible. Notice the use of the support v4 and v7 Google libraries. These allow this code to run on a Lollipop (or lower) device and act just the same as before.
+
+    ### Conclusion
+
+    Granted we now have to write more code for our apps when it comes to permissions on Marshmallow, but I think you can also see the benefits. Users will be much happier granting permissions, they are much more likely to install your application without a wall of permissions at the start. You as a developer can degrade your app gracefully so that when users deny permissions your app still works. (Not granted the Camera permission? See if the user wants to select a photo from Facebook.)
```

Comment:
`You as a developer can degrade your app` -> `You, as a developer, can degrade your app`